internal/source/image.go

```diff
@@ -0,0 +1,188 @@
+		pod.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{{Name: "util", MountPath: "/bin"}}
+		return nil
+	})
+}
+
+func pendingPodResult(pod *corev1.Pod) *Result {
+	var messages []string
+	for _, cStatus := range append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...) {
+		if cStatus.State.Waiting != nil && cStatus.State.Waiting.Reason == "ErrImagePull" {
+			messages = append(messages, cStatus.State.Waiting.Message)
+		}
+		if cStatus.State.Waiting != nil && cStatus.State.Waiting.Reason == "ImagePullBackOff" {
+			messages = append(messages, cStatus.State.Waiting.Message)
+		}
+	}
+	return &Result{State: StatePending, Message: strings.Join(messages, "; ")}
+}
+
+func (i *Image) failedPodResult(ctx context.Context, pod *corev1.Pod) error {
+	logs, err := i.getPodLogs(ctx, pod)
+	if err != nil {
+		return fmt.Errorf("unpack failed: failed to retrieve failed pod logs: %w", err)
+	}
+	_ = i.Client.Delete(ctx, pod)
+	return fmt.Errorf("unpack failed: %v", string(logs))
+}
+
+func (i *Image) succeededPodResult(ctx context.Context, pod *corev1.Pod) (*Result, error) {
+	bundleFS, err := i.getBundleContents(ctx, pod)
+	if err != nil {
+		return nil, fmt.Errorf("get bundle contents: %w", err)
+	}
+
+	digest, err := i.getBundleImageDigest(pod)
+	if err != nil {
+		return nil, fmt.Errorf("get bundle image digest: %w", err)
+	}
+
+	resolvedSource := &rukpakv1alpha1.BundleSource{
+		Type:  rukpakv1alpha1.BundleSourceTypeImage,
+		Image: &rukpakv1alpha1.ImageSource{Ref: digest},
+	}
+
+	return &Result{Bundle: bundleFS, ResolvedSource: resolvedSource, State: StateUnpacked}, nil
+}
+
+func (i *Image) getBundleContents(ctx context.Context, pod *corev1.Pod) (fs.FS, error) {
+	bundleData, err := i.getPodLogs(ctx, pod)
+	if err != nil {
+		return nil, fmt.Errorf("get bundle contents: %w", err)
+	}
+	bd := struct {
+		Content []byte `json:"content"`
+	}{}
+
+	if err := json.Unmarshal(bundleData, &bd); err != nil {
+		return nil, fmt.Errorf("parse bundle data: %w", err)
+	}
+
+	gzr, err := gzip.NewReader(bytes.NewReader(bd.Content))
+	if err != nil {
+		return nil, fmt.Errorf("read bundle content gzip: %w", err)
+	}
+	return tarfs.New(gzr)
+}
+
+func (i *Image) getBundleImageDigest(pod *corev1.Pod) (string, error) {
+	for _, ps := range pod.Status.ContainerStatuses {
+		if ps.Name == bundleUnpackContainerName && ps.ImageID != "" {
+			return ps.ImageID, nil
+		}
+	}
+	return "", fmt.Errorf("bundle image digest not found")
+}
+
+func (i *Image) getPodLogs(ctx context.Context, pod *corev1.Pod) ([]byte, error) {
+	logReader, err := i.KubeClient.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &corev1.PodLogOptions{}).Stream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer logReader.Close()
+	buf := &bytes.Buffer{}
+	if _, err := io.Copy(buf, logReader); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func (i *Image) handleUnexpectedPod(ctx context.Context, pod *corev1.Pod) error {
```

Comment:
nit: DI the client and make this a function?

Reply:
Main reason I didn't do that is because I expect this package to contain at least a handful of other sources and I wanted to keep helper functions scoped to the specific source they're for.

I can change it to `handleUnexpectedImageUnpackPod` though?

Reply:
Okay that's fine with me. Let's try and land the core functionality first, so I don't think we need to rename this method right now.